src/fsipy/automatedPostprocessing/postprocessing_mesh/separate_domain.py

```diff
@@ -0,0 +1,143 @@
+    mesh = Mesh()
+    with HDF5File(mesh.mpi_comm(), str(mesh_path), "r") as hdf:
+        hdf.read(mesh, "/mesh", False)
+        domains = MeshFunction("size_t", mesh, mesh.topology().dim())
+        hdf.read(domains, "/domains")
+        boundaries = MeshFunction("size_t", mesh, mesh.topology().dim() - 1)
+        hdf.read(boundaries, "/boundaries")
+
+    for domain_id, domain_name in zip([fluid_domain_id, solid_domain_id], ["fluid", "solid"]):
+
+        domain_of_interest = SubMesh(mesh, domains, domain_id)
+        domain_of_interest_path = mesh_path.with_name(mesh_path.stem + f"_{domain_name}.h5")
+        print(f" --- Saving {domain_name} domain to {domain_of_interest_path} \n")
+        with HDF5File(domain_of_interest.mpi_comm(), str(domain_of_interest_path), "w") as hdf:
+            hdf.write(domain_of_interest, "/mesh")
+
+        # Save for viewing in paraview
+        if view:
+            domain_of_interest_pvd_path = domain_of_interest_path.with_suffix(".pvd")
+            File(str(domain_of_interest_pvd_path)) << domain_of_interest
+
+    print(" --- Done separating domains \n")
+
+    with h5py.File(mesh_path) as vectorData:
+        domain_values = vectorData['domains/values'][:]
+        domain_topology = vectorData['domains/topology'][:, :]
+        domain_coordinates = vectorData['mesh/coordinates'][:, :]
+
+    for domain_id, domain_name in zip([fluid_domain_id, solid_domain_id], ["fluid", "solid"]):
+        # non-zero is used to find the indices of the domain of interest
+        domain_of_interest_index = (domain_values == domain_id).nonzero()
+        # extract the topology of the domain of interest
+        domain_of_interest_topology = domain_topology[domain_of_interest_index[0], :]
+        # Here, we want to extract the coordinates of the domain of interest
+        # We can do this by extracting the unique node IDs of the domain of interest from the topology
+        domain_of_interest_ids = np.unique(domain_of_interest_topology)  # unique will return sorted array
+        domain_of_interest_coordinates = domain_coordinates[domain_of_interest_ids, :]
+        # Fix topology of the each domain
+        # This is necessary because the node numbering may not be continuous
+        # while there is one to one correspondence between the node IDs and the coordinates
+        print(f" --- Fixing topology of {domain_name} domain \n")
+        print(" --- This is computationally expensive \n")
+        if not np.all(np.diff(domain_of_interest_ids) == 1):
+            for node_id, domain_id in enumerate(domain_of_interest_ids):
+                if node_id % 1000 == 0:
+                    print(f" --- Fixing topology of {domain_name} domain: {node_id} / {len(domain_of_interest_ids)} \n")
+                mask = domain_of_interest_topology == domain_id
+                domain_of_interest_topology[mask] = node_id
+        else:
+            print(f" --- {domain_name} topology does not need to be fixed \n")
```

Comment:
@keiyamamo could you drop by at some point so we can discuss this?

Reply:
@jorgensd  Of course!